## 프로그램 오류의 종류

### 컴파일 에러
컴파일 시 발생하는 에러
### 런타임 에러
실행할 때 발생하는 에러
### 논리적 에러
실행시에 의도와 다르게 동장하는 에러

## 예외 처리 구문 - try-catch-finally
```java
try{
	//예외가 발생할 가능성이 있는 문장을 넣는다.
} catch(Exception1 e1){
	//Exception1이 발생할 경우, 이를 처리하기 위한 문장을 적는다. 
} finally{
	// 예외의 발생여부에 관계없이 항상 수행되어야 하는 문
}
```

## 예외 발생시키기
1. 연산자 new를 이용하여 발생시키려는 예외 클래스의 객체를 만든다.
	`Exception e = new Exception("message");`
2. 키워드 throw를 이용해서 예외 발생
	`throw e;`

## 예외 클래스의 계층 구조
![[Pasted image 20230918191231.png]]

try 블럭에서 예외가 발생하면, 발생한 예외를 처리할 catch 블럭을 첫번째 블럭부터 순서대로 찾는다. 예외의 최고 조상인 Exception을 처리하는 catch 블럭은 모든 종류의 예외를 처리할 수 있다. (반드시 마지막 catch 블럭이어야 한다.)
> 상속 관계에 있는 예외를 확인하고 더 작은 단위의 예외를 우선순위가 높은 catch블럭으로 만들어야 한다! 


	 발생한 예외 객체를 catch블럭의 참조변수로 접근할 수 있다.
[]![[Pasted image 20230918192310.png]]

